### ISSUE TITLE
Use list of approved domains instead of ignored ones

### ISSUE THREAD

**Issue by frvge** [webhintio/hint#766]
Instead of ignoring domains, which can be tricky if you have ad networks, I'd like to propose a new feature that only checks a resource if its domain matches a regex.

This way, I can allow cdn.site.com, www.site.de and a few other variants (preferably in 1 regex) and ignore all the other domains.

See https://github.com/webhintio/hint/issues/766#issuecomment-379814475 for more details about implementation.

**Comment by molant** [comment]
@frvge good idea. Any proposal for the name of the property?
Also, do you want to try to implement it? We'll be more than happy to help you in the process.

**Comment by alrra** [comment]
The blacklist approach makes users actually acknowledge the issues, and only after, decide if they want to ignore certain domains or try to do something about them.

With the whitelist approach the biggest downside is that over time the site may include new things, users will most probably forget to update the list, and while `sonarwhal` may suggest that everything is fine, the site as a whole may be getting worst.

---

Note: I'm fine with having this feature, but we should make the advantages and disadvantages of the approaches perfectly clear.

**Comment by frvge** [comment]
@molant , I don't have experience with TypeScript, or anything after ES3, but I can probably get some assistance from my colleagues. I'll give it a try.

@alrra , correct. There is a risk of missing out on issues. This should be worded clearly in the documentation.

**Comment by alrra** [comment]
>  I don't have experience with TypeScript, or anything after ES3, but I can probably get some assistance from my colleagues. I'll give it a try.

If there is something we can help with, let us know!

**Comment by frvge** [comment]
I started working on it today, but it's too early to really say if I'll be able to get it done.

**Comment by molant** [comment]
Awesome!

**Comment by alrra** [comment]
@frvge Let us know if we can help with anything!

**Comment by frvge** [comment]
Do you want to have an additional parameter, just like the ignoredUrls one? Or maybe a setting within the ignoredUrls like "reverse"? Or maybe it's better to abstract it away in a new structure / interface?

Currently I went with the additional parameter, but I'm doubting if it's the cleanest way.

The idea is that if the whitelist has at least 1 element and a resource is not in the whitelist, it gets ignored. Afterwards, the ignore list gets applied.

I'm not sure if I want to have a "rules" part on the whitelist because it doesn't make sense to whitelist a domain/resource for only a few rules. That functionality is better done by ignoreUrls.

**Comment by molant** [comment]
Hi @frvge 

I think an additional parameter is good. What name are you using right now? For my examples I'll use `whitelist` but don't have any opinion at the moment.

From what I read I think we are on the same page. Let me write it down just to make sure:

| `whiteList` | `ignoredUrls` | What happens |
| ----------- | ------------- | ---------------- | 
|      ❌       |        ✔        | Same behavior as now | 
|      ✔       |       ❌         | Only the URLs in the `whitelist` are processed with the configured rules |
|     ✔        |       ✔         | Only the URLs in the `whitelist` are processed with the configured rules <br/> If `ignoredUrls` contains URLs that are in `whitelist`, we apply the rule restrictions to those urls |

Using the following configuration:

```js
{
  whitelist: ['mydomain.com', 'cdnmydomain.com'],
  ignoredUrls: [{
    domain: 'cdnmydomain.com',
    rules: ['no-disallowed-headers']
  }]
}
```

This means that `sonarwhal` will only analyze the resources coming from `mydomain.com` and `cdnmydomain.com`, and for those coming from `cdnmydomain.com` the rule `no-disallowed-headers` will not be executed.

Taking into account this, someone could have a configuration like the following:

```js
{
  whitelist: ['mydomain.com', 'cdnmydomain.com'],
  ignoredUrls: [{
    domain: 'cdnmydomain.com',
    rules: ['*']
  }]
}
```
which would be the same as:

```js
{
  whitelist: ['mydomain.com'] 
}
```

So basically `whitelist` takes precedence over `ignoredUrls`, and `ignoredUrls` only gets applied if some of the urls match:

1. filter first by regular expressions in `whitelist`
2. filter then again by urls in `ignoredUrls`

Is this what you were thinking? Did I miss something or any edge case?

Thanks for taking the time and working on this!

**Comment by frvge** [comment]
Yes, this is correct. Thanks for the clarifying comment. I want to add that I appreciate the Architecture diagram. The documentation is very good.

**Comment by frvge** [comment]
@molant , my priorities have shifted for the moment. You (aka the members) can continue with my branch, or leave it as is, or remove my assignment from this task. Once I'm done, I'll come back to this. I just don't know when that'll be.
